default/python/universe_generation/empires.py

```diff
@@ -128,36 +271,47 @@ def compile_home_system_list(num_home_systems, systems):
-              % (len(systems), num_home_systems, min_jumps)
-        # ...no, decrease the min jump distance and try again
-        min_jumps -= 1
 
-    # check if the loop above delivered a list with enough home systems, or if it exited because the min jump distance
-    # has been decreased to 0 without finding enough systems
-    # in that case, our galaxy obviously is too crowded, report an error and return an empty list
+    # home systems must have a certain minimum of systems in their near vicinity
+    # we will try to select our home systems from systems that match this criteria, if that fails, we will select our
+    # home systems from all systems and add the missing number planets to the systems in their vicinity afterwards
+    # the minimum planet limit and the jump range that defines the "near vicinity" are controlled by the
+    # HS_* option constants in options.py (see there)
+
+    # lets start by filtering out all systems from the pool we got passed into this function that match the criteria
+    filtered_pool = [s for s in systems if has_min_planets_in_vicinity(s)]
+    print "Filtering out systems that meet the minimum planets in the near vicinity condition yielded",\
+        len(filtered_pool), "systems"
+    print "Using this as the preferred pool for home system selection"
+    # now try to pick the requested number of home systems by calling find_home_systems
+    # this function takes two pools, a "complete" pool and one with preferred systems
+    # it will try to pick the home systems from the preferred pool first, so pass our filtered pool as preferred pool
+    home_systems = find_home_systems(num_home_systems, systems, filtered_pool, min_jumps)
+
+    # check if the selection process delivered a list with enough home systems
+    # if not, our galaxy obviously is too crowded, report an error and return an empty list
     if len(home_systems) < num_home_systems:
-        util.report_error("Python generate_home_system_list: requested %d homeworlds in a galaxy with %d systems"
-                          % (num_home_systems, len(systems)))
+        report_error("Python generate_home_system_list: requested %d homeworlds in a galaxy with %d systems"
+                     % (num_home_systems, len(systems)))
         return []
 
+    # check if we got more home systems than we requested
+    if len(home_systems) > num_home_systems:
+        # yes: calculate the number of planets in the near vicinity of each system
+        # and store that value with each system in a map
+        hs_planets_in_vicinity_map = {s: count_planets_in_systems(get_systems_within_jumps(s, HS_VICINITY_RANGE))
+                                      for s in home_systems}
+        # sort the home systems by the number of planets in their near vicinity using the map
+        # now only pick the number of home systems we need, taking those with the highest number of planets
+        home_systems =\
```

Comment:
PyCharm  usually don`t change number of lines. Reformating uses existing linebreaks. It will produce exacly same result if you break line before `reverse`

`key` is function that accepts single argument.
You can use: `key=hs_planets_in_vicinity_map.__getitem__` or `key=hs_planets_in_vicinity_map.get` to avoid lambda.  (and after that it will fit single line)